Sources/OpenAPIURLSession/URLSessionBidirectionalStreaming/BidirectionalStreamingURLSessionDelegate.swift

```diff
@@ -0,0 +1,162 @@
+        self.requestBody = requestBody
+        self.hasAlreadyIteratedRequestBody = false
+        self.hasSuspendedURLSessionTask = false
+        self.requestStreamBufferSize = requestStreamBufferSize
+        (self.responseBodyStream, self.responseBodyStreamSource) = AsyncBackpressuredStream.makeStream(
+            backPressureStrategy: .highLowWatermarkWithElementCounts(
+                lowWatermark: responseStreamWatermarks.low,
+                highWatermark: responseStreamWatermarks.high
+            )
+        )
+    }
+
+    func urlSession(_ session: URLSession, needNewBodyStreamForTask task: URLSessionTask) async -> InputStream? {
+        debug("Task delegate: needNewBodyStreamForTask")
+        // If the HTTP body cannot be iterated multiple times then bad luck; the only thing
+        // we can do is cancel the task and return nil.
+        if hasAlreadyIteratedRequestBody {
+            guard requestBody!.iterationBehavior == .multiple else {
+                debug("Task delegate: Cannot rewind request body, canceling task")
+                task.cancel()
+                return nil
+            }
+        }
+
+        // Create a fresh pair of streams.
+        let (inputStream, outputStream) = createStreamPair(withBufferSize: requestStreamBufferSize)
+
+        // Bridge the output stream to the request body and open the output stream.
+        requestStream = HTTPBodyOutputStreamBridge(outputStream, requestBody!, openOutputStream: true)
+
+        // Return the new input stream (unopened, it gets opened by URLSession).
+        return inputStream
+    }
+
+    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+        debug("Task delegate: didReceive data (numBytes: \(data.count), dataTask.state: \(dataTask.state))")
+        do {
+            switch try responseBodyStreamSource.write(contentsOf: [ArraySlice(data)]) {
+            case .produceMore: break
+            case .enqueueCallback(let writeToken):
+                if self.hasSuspendedURLSessionTask {
+                    debug("Task delegate: already suspended task, not enqueing another writer callback")
+                    break
+                }
+                debug("Task delegate: response stream backpressure, suspending URLSession task and enqueing callback")
+                dataTask.suspend()
+                self.hasSuspendedURLSessionTask = true
+                responseBodyStreamSource.enqueueCallback(writeToken: writeToken) { result in
+                    switch result {
+                    case .success:
+                        debug("Task delegate: response stream callback, resuming URLSession task")
+                        dataTask.resume()
+                        self.hasSuspendedURLSessionTask = false
+                    case .failure(let error):
+                        debug("Task delegate: response stream callback, cancelling URLSession task, error: \(error)")
```

Comment:
Will we surface this error to the adopter?

Reply:
It will surface at the HTTP request failed and the `ClientError.underlyingError` will have a `URLError` with operation cancelled IIUC.

We have the following tests:
- `testHTTPBodyOutputStreamBridgeStreamClosedImmediately`
- `testHTTPBodyOutputStreamBridgeStreamClosedEarly`

But we don't have a test at the URLSession level here. 

I can write a test that's independent of all this logic to show what surface in the case the URLSessionTask gets cancelled if you like? Such a test would provide a session delegate that always cancels the task and we can see what error comes out of the `URLSessionTransport`.

Reply:
Doing a one-off test is also fine, I'm just curious whether the adopter gets enough information to start debugging the failure; that we're not dropping any diagnostics here.

Reply:
Added a test for this. `send` fails with `URLError` with `errorCode == .cancelled`. This should end up further up the stack as a `ClientError`.

Do we think this is enough, or would we like an explicit (name TDB) `URLSessionTransportError.cannotReplayResponseBody`?

Also, this interplays with the discussion on whether we do something smarter when given a request body with single iteration behavior.

Reply:
What does the `error` hold that is being provided by `case .failure(let error):`? Is that relevant to the user?

Reply:
This is coming from the `AsyncBackpressuredStream` and IIUC (@FranzBusch can confirm) it means the consumer went away, i.e. the iterator went out of scope. I _think_ therefore, it means they didn't finish consuming the response body for some reason. Presumably, in this case, they're not interested in the error by that time? 

It gets returned by the stream if the `action` after enqueuing the producer is `resumeProducerWithCancellationError`, and it always returns `CancellationError()`:

```swift
            case .resumeProducerWithCancellationError(let onProduceMore): onProduceMore(.failure(CancellationError()))
````

Reply:
Cool, yeah then we can just always represent this by a cancelation error.